Update MQTT 5 uTransport specification (#135)
eclipse-uprotocol/up-spec: up-l1/mqtt_5.adoc

= MQTT 5 Transport Protocol
:toc:
:sectnums:

The key words "*MUST*", "*MUST NOT*", "*REQUIRED*", "*SHALL*", "*SHALL NOT*", "*SHOULD*", "*SHOULD NOT*", "*RECOMMENDED*", "*MAY*", and "*OPTIONAL*" in this document are to be interpreted as described in https://www.rfc-editor.org/info/bcp14[IETF BCP14 (RFC2119 & RFC8174)]

----

> SPDX-FileCopyrightText: 2023 Contributors to the Eclipse Foundation
> 
> Licensed to the Apache Software Foundation (ASF) under one
> or more contributor license agreements.  See the NOTICE file
> distributed with this work for additional information
> regarding copyright ownership.  The ASF licenses this file
> to you under the Apache License, Version 2.0 (the
> "License"); you may not use this file except in compliance
> with the License.  You may obtain a copy of the License at
> 
>   http://www.apache.org/licenses/LICENSE-2.0
> 
> Unless required by applicable law or agreed to in writing,
> software distributed under the License is distributed on an
> "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
> KIND, either express or implied.  See the License for the
> specific language governing permissions and limitations
> under the License.
> 
> SPDX-FileType: DOCUMENTATION
> SPDX-License-Identifier: Apache-2.0

Reviewer: ```suggestion
SPDX-FileCopyrightText: 2023 Contributors to the Eclipse Foundation

See the NOTICE file(s) distributed with this work for additional
information regarding copyright ownership.

This program and the accompanying materials are made available under
the terms of the Apache License Version 2.0 which is available at
https://www.apache.org/licenses/LICENSE-2.0
 
SPDX-FileType: DOCUMENTATION
SPDX-License-Identifier: Apache-2.0
```

Author: updated

For more information, please refer to https://mqtt.org/

This document will discuss the uTransport implementation on MQTT 5.

== Specifications

=== UAttribute Mapping

MQTT 5 supports custom header fields, and we leverage this to map the UAttribute values into the MQTT header.

[cols="1,1"]
|===
| MQTT Header Property Code | Value (key:value)
| UserProperty
| ("1", "{UAttribute.id}") # UUID id
| UserProperty
| ("2", "{UAttribute.type}") # UMessageType
| UserProperty
| ("3","{UAttribute.source}") # Short Form Serialized Source UURI
| UserProperty
| ("4","{UAttribute.sink}") # Short Form Serialized Sink UURI
| UserProperty
| ("5","{UAttribute.priority}") # UPriority
| UserProperty
| ("6", "{UAttribute.ttl}") # Optional ttl
| UserProperty
| ("7", "{UAttribute.permissionLevel}") # Optional permission level
| UserProperty
| ("8", "{UAttribute.commStatus}") # Optional UCode commstatus
| UserProperty
| ("9", "{UAttribute.reqId}") # UUID reqid
| UserProperty
| ("10", "{UAttribute.token}") # Optional token
| UserProperty
| ("11", "{UAttribute.traceParent}") # Optional traceparent

> |===

Author: Add encoding information for each userProperty

> |===

Author: added

> All uAttributes are mapped to a MQTT header UserProperty, where the key is the UAttribute protobuf field number. The value is a string representation of the UAttribute field.

Author: Add line about optional headers not being included as userProperty if not used

Author: updated

Has sink UUri:

`{client_identifier}/{source.authority_name}/{sink.authority_name}/{sink.ue_id}/{sink.ue_version_major}/{sink.resource_id}`

No sink UUri (Publish message type only):

`{client_identifier}/{source.authority_name}//{source.ue_id}/{source.ue_version_major}/{source.resource_id}`

The entity and resource elements always represent the destination of the message. In the case of a Publish message type, the destination happens to be based on the source UUri. The source authority information is also always present to allow for authorization controls based on the authority information, since an authority maps to the uE sending the message.

The client_identifier is used to determine where messages are coming from. The valid values for the client_identifier are:

- `c` : Indicates that the message originated from the cloud
- `d` : Indicates that the message originated from a device and is remote

> - `l` : Indicates that the message originated from a device and is local to that device

Author: Drop `L` for now. We can add back later if needed

Author: removed

> The client_identifier that is used for the topic is based on the kind of MQTT uTransport client that is sending the message. An MQTT uTransport client can be one of the following:

Reviewer 2: Hey @devkelley -- following up on this from our talk this morning.

Is the client_identifier something that would be set in, say, the constructor of `UPClientMQTT5`? I'm trying to understand where this information comes from so that when an MQTT5 topic name is crafted within `UPClientMQTT5`, the correct client_identifier is prepended.

Author: Correct, this would be initialized by the constructor of the UPClientMQTT5. If a client was initialized on a device then `d` would be prepended, and if a client was initialized in the cloud then `c` would be prepended.

- Cloud MQTT uTransport client: A client that runs in the cloud.
- Remote MQTT uTransport client: A client that runs on a device and is connected to a cloud MQTT broker.
- Local MQTT uTransport client: A client that runs on a device and is connected to a local MQTT broker.

Below are examples of messages being sent and how to subscribe to those messages:

==== Remote uTransport Client Examples

How a source and sink UUri are mapped to an MQTT topic when the message is sent from a remote uTransport client:

[cols="1,1,1,1"]
|===
| source | sink | send topic | listen topic

> | /vehicle_1.veh.com/1234/1/5678 | /vehicle_2.veh.com/4321/1/8765 | d/vehicle_1.veh.com/vehicle_2.veh.com/4321/1/8765 | c/vehicle_1.veh.com/vehicle_2.veh.com/4321/1/8765

Author: updated the examples to make it clear its base16 encoded

| /vehicle_1.veh.com/1234/1/5678 | None | d/vehicle_1.veh.com//1234/1/5678 | c/vehicle_1.veh.com//1234/1/5678/
|===

Listener examples with wildcards:

[cols="1,1,1,1"]
|===
| goal | source filter | sink filter | listener topic

> | Subscribe to all incoming message to a UAuthority | empty | /vehicle_1.veh.com/FFFFFFFF/FF/FFFF | c/\+/vehicle_1.veh.com/+/\+/+

Reviewer: The sink filter needs to be
`//vehicle_1.veh.com/FFFF/FF/FFFF`
in this case. The service instance ID wildcard is 0x0000.

> | Subscribe to all publish messages from a UAuthority | /vehicle_1.veh.com/FFFFFFFF/FF/FFFF | None | c/vehicle_1.veh.com//\+/+/+

Reviewer: Source filter needs to be
`//vehicle_1.veh.com/FFFF/FF/FFFF`

Author: updated

|===

==== Cloud uTransport Client Examples

Shows how source and sink filters can be used to construct wildcard topics for listening to messages from many devices.

[cols="1,1,1,1"]
|===
| goal | source filter | sink filter | listener topic
| Subscribe to all publish messages from devices | empty | None | d/\+//+/\+/+

> | Subscribe to all messages sent from a UAuthority | /vehicle_1.veh.com/FFFFFFFF/FF/FFFF | empty | d/vehicle_1.veh.com/\+/+/\+/+

Reviewer: Source filter needs to be
`//vehicle_1.veh.com/FFFF/FF/FFFF`

Author: updated

See the NOTICE file(s) distributed with this work for additional
information regarding copyright ownership.

This program and the accompanying materials are made available under
the terms of the Apache License Version 2.0 which is available at
https://www.apache.org/licenses/LICENSE-2.0

SPDX-FileType: DOCUMENTATION
SPDX-License-Identifier: Apache-2.0
----

== Overview

MQTT is an OASIS standard messaging protocol for the Internet of Things (IoT). It is designed as an extremely lightweight publish/subscribe messaging transport that is ideal for connecting remote devices with a small code footprint and minimal network bandwidth. MQTT today is used in a wide variety of industries, such as automotive, manufacturing, telecommunications, oil and gas, etc

For more information, please refer to https://mqtt.org/

This document will discuss the uTransport implementation on MQTT 5.

== Specifications

=== UAttribute Mapping

MQTT 5 supports custom header fields, and we leverage this to map the UAttribute values into the MQTT header.

[cols="1,1,1,1"]
|===
| MQTT Header Property Code | Value (key:value) | Value Type | Description

| UserProperty

> | ("0", "{UAttribute Version}")

Reviewer 3: ```suggestion
| ("0", "{UAttributes Version}")
```

| String
a| The version of the UAttribute protobuf

* MUST represent the major version of the UAttribute protobuf. Currently should be set to `1`.

| UserProperty

> | ("1", "{UAttribute.id}")

Reviewer 3: ```suggestion
| ("1", "{UAttributes.id}")
```

| String
a| Unique identifier for the message

* MUST be set to the https://www.rfc-editor.org/rfc/rfc4122.html#section-3[hyphenated string representation] of the UUID.

| UserProperty

> | ("2", "{UAttribute.type}")

Reviewer 3: ```suggestion
| ("2", "{UAttributes.type}")
```

| String
a| The type of message

* MUST be the enum value for UAttributes.type link:../up-core-api/uprotocol/uattributes.proto[UMessageType enum].

| UserProperty

> | ("3","{UAttribute.source}")

Reviewer 3: ```suggestion
| ("3","{UAttributes.source}")
```

| String
a| The origin (address) of the message

* MUST be the link:../basics/uri.adoc[string serialization of the UUri]

| UserProperty
| ("4","{UAttribute.sink}")
| String
a| The destination (address) of the message

* MUST be the link:../basics/uri.adoc[string serialization of the UUri]

| UserProperty

> | ("5","{UAttribute.priority}")

Reviewer 3: ```suggestion
| ("5","{UAttributes.priority}")
```

| String
a| The message's priority as defined in link:../basics/qos.adoc[QoS doc]

* MUST be set to the enum value for UAttributes.priority link:../up-core-api/uprotocol/uattributes.proto[UPriority enum].

| UserProperty

> | ("6", "{UAttribute.ttl}")

Reviewer 3: ```suggestion
| ("6", "{UAttributes.ttl}")
```

| String
a| The amount of time (in milliseconds) after which this message MUST NOT be delivered/processed anymore

| UserProperty

> | ("7", "{UAttribute.permissionLevel}")

Reviewer 3: ```suggestion
| ("7", "{UAttributes.permissionLevel}")
```

| String
a| The service consumer's permission level as defined in link:../up-l2/permissions.adoc#_code_based_access_permissions_caps[Code-Based uE Access Permissions (CAPs)]

| UserProperty

> | ("8", "{UAttribute.commStatus}")

Reviewer 3: ```suggestion
| ("8", "{UAttributes.commStatus}")
```

| String
a| A UCode indicating an error that has occurred during the delivery of either an RPC Request or Response message

* MUST be set to the link:../up-core-api/uprotocol/ustatus.proto[UCode]'s integer value as a string.

| UserProperty

> | ("9", "{UAttribute.reqId}")

Reviewer 3: ```suggestion
| ("9", "{UAttributes.reqId}")
```

| String
a| The identifier that a service consumer can use to correlate an RPC Repsonse message with its RPC Request

* MUST be set to the https://www.rfc-editor.org/rfc/rfc4122.html#section-3[hyphenated string representation] of the UUID.

| UserProperty

> | ("10", "{UAttribute.token}")

Reviewer 3: ```suggestion
| ("10", "{UAttributes.token}")
```

| String
a| The service consumer's access token

| UserProperty

> | ("11", "{UAttribute.traceParent}")

Reviewer 3: ```suggestion
| ("11", "{UAttributes.traceParent}")
```

| String
a| A tracing identifier to use for correlating messages across the system

| UserProperty

> | ("upf", "{UPayload.format}")

Reviewer 3: ```suggestion
| ("12", "{UAttributes.UPayloadFormat}")
```

Author: Updated to UAttributes.payload_format to match rest of table.

The client_identifier is used to determine where messages are coming from. The valid values for the client_identifier are:

- `c` : Indicates that the message originated from the cloud
- `d` : Indicates that the message originated from a device and is remote

The client_identifier that is used for the topic is based on the kind of MQTT uTransport client that is sending the message. An MQTT uTransport client can be one of the following:

- Cloud MQTT uTransport client: A client that runs in the cloud.
- Remote MQTT uTransport client: A client that runs on a device and is connected to a cloud MQTT broker.

Below are examples of messages being sent and how to subscribe to those messages:

==== Remote uTransport Client Examples

How a source and sink UUri are mapped to an MQTT topic when the message is sent from a remote uTransport client:

[cols="1,1,1,1"]
|===

> | source | sink | send topic | listen topic

Author: Good catch, I do not think we do. Will remove when I update the topic structure

> | source | sink | send topic | listen topic

Author: now a single column

For more information, please refer to https://mqtt.org/

This document will discuss the uTransport implementation on MQTT 5.

== Specifications

=== UAttribute Mapping

MQTT 5 supports custom header fields, and we leverage this to map the UAttribute values into the MQTT header. Only required uAttribute values are mapped into their own UserProperty fields.

[cols="1,1"]
|===
| MQTT Header Property Code | Value (key:value)
| UserProperty
| ("1", "{UAttribute.id}") # UUID id
| UserProperty
| ("2", "{UAttribute.type}") # UMessageType
| UserProperty
| ("3","{UAttribute.source}") # Short Form Serialized Source UURI
| UserProperty
| ("4","{UAttribute.sink}") # Short Form Serialized Sink UURI
| UserProperty
| ("5","{UAttribute.priority}") # UPriority
| UserProperty
| ("9", "{UAttribute.reqId}") # UUID reqid
| UserProperty

> | ("opt_ua", "{byte array of remaining UAttributes}") # Remaining UAttributes

Reviewer 3: @devkelley why don't we just add the optional attributes the same way as the mandatory (as additional key:value pairs?

Author: After our weekly mqtt discussion, this has been reverted back so that each UAttribute has its own key value pair

|===

All required uAttributes are mapped to a MQTT header UserProperty, where the key is the UAttribute protobuf field number.

The source and sink fields **MUST** be a short form serialized UUri.

Optional uAttributes will be a string representation of a byte array under the UserProperty field with key "opt_ua". This is done to reduce the size of the overall payload. Protobuf serialized byte arrays do not include values in a protobuf message object that are not set, so the byte array will only include the optional UAttributes that are set.

=== URI Mapping to MQTT topics

The MQTT topic a message is published on utilizes elements from both the source and sink UUri fields. Which elements from the UUri are used in the topic depend on whether the sink UUri is present.

Has sink UUri:

> `{source.authority.domain}/{source.authority.name}/{sink.authority.domain + ":" + sink.authority.name}/{sink.entity.id}/{sink.entity.version_major}/{sink.resource.id}`

Reviewer 3: Not sure I follow the addition of ":" but we can talk about it in a few hours

Author: Removed with update to match topic naming to 1.5.8 UURI spec

> `{source.authority.domain}/{source.authority.name}/{sink.authority.domain + ":" + sink.authority.name}/{sink.entity.id}/{sink.entity.version_major}/{sink.resource.id}`

Reviewer 3: Is there any added benefit if the domain is in reverse-DNS notation?

Author: Removed with update to match topic naming to 1.5.8 UURI spec

No sink UUri (Publish message type only):

`{source.authority.domain}/{source.authority.name}//{source.entity.id}/{source.entity.version_major}/{source.resource.id}`

The entity and resource elements always represent the destination of the message. In the case of a Publish message type, the destination happens to be based on the source UUri. The source authority information is also always present to allow for authorization controls based on the authority information, since an authority maps to the uE sending the message. The domain information is included is separated out as well to allow for the cloud to easily subscribe to messages from devices only.

Below are examples of messages being sent and how to subscribe to those messages:

==== Example 1: Notification or RPC Message Types

> [source]

Reviewer 3: This needs to be updated to refect the latest version of UUri that no longer has UAuthority, UEntity, UResource and is much simplified. Also you can use the short string form representation since that is all that we use now (there is no other form)

Author: Done

----
source:
UUri {
  authority: UAuthority {
    name: "Vehicle_1",
    domain: veh.com,
  },
  entity: UEntity {
    version_major: 1,
    id: 1234,
  },
  resource: UResource {
    id: 5678,
  }
}

sink:
UUri {
  authority: UAuthority {
    name: "Vehicle_2",
    domain: veh.com,
  },
  entity: UEntity {
    version_major: 1,
    id: 4321,
  },
  resource: UResource {
    id: 8765,
  }
}
----
topic messages are published to: `veh.com/Vehicle_1/veh.com:Vehicle_2/4321/1/8765`

> to register a listener with this topic, the following source and sink filter would need to be provided:

Reviewer 3: Use the short form string now to represent the URI

Author: Updated

    id: 5678,
  }
}
----

==== Example 3: Subscribe to all incoming messages to a UAuthority

[source]
----
source filter: empty

sink filter:
UUri {
  authority: UAuthority {
    name: "Vehicle_1",
    domain: veh.com,
  },
  entity: empty,
  resource: empty,
}
----

topic: `+/+/veh.com:Vehicle_1/+/+/+`

==== Example 4: Subscribe to all publish messages from a UAuthority

[source]
----
source filter:
UUri {
  authority: UAuthority {
    name: "Vehicle_1",
    domain: veh.com,
  },
  entity: empty,
  resource: empty,
}

sink filter: None
----

topic: `veh.com/Vehicle_1//+/+/+`

==== Example 5: Subscribe to all publish messages from a domain

[source]
----
source filter:
UUri {
  authority: UAuthority {
    name: empty,
    domain: veh.com,
  },
  entity: empty,
  resource: empty,
}

sink filter: None
----

topic: `veh.com/+//+/+/+`

==== Example 6: Subscribe to all messages sent from a UAuthority

This would include all message types.

[source]
----
source filter:
UUri {
  authority: UAuthority {
    name: "Vehicle_1",
    domain: veh.com,
  },
  entity: empty,
  resource: empty,
}

sink filter: empty
----

topic: `veh.com/Vehicle_1/+/+/+/+`

=== Payload Encoding

> The MQTT payload **MUST** be a UPayload that is represented as a byte array to reduce size.

Reviewer 3: Does the payload need to be encoded in a specific format (base64 encoded) or can it be binary?

Author: That is an excellent question. I believe that base64 encoding is used in the rust implementation. But I am not 100% sure: https://docs.rs/protobuf/latest/protobuf/trait.Message.html#method.write_to_bytes. @PLeVasseur do you happen to know if this is for sure base 64 encoding?

Reviewer: The payload of an MQTT message is just an array of bytes. No need to base64 encode it.

Author: We will pass the upayload.format in the userProperty header ("upayloadformat", ENUM). upayload.data will be passed in the message as bytes.

Author: Added to header definition

== Specifications

=== UAttributes Mapping

MQTT 5 supports custom header fields, and we leverage this to map the UAttributes values into the MQTT header.

[cols="1,1,1,1"]
|===
| MQTT Header Property Code | Value (key:value) | Value Type | Description

| UserProperty
| ("0", "{UAttributes Version}")
| String
a| The version of the UAttributes protobuf

* MUST represent the major version of the UAttributes protobuf. Currently should be set to `1`.

| UserProperty
| ("1", "{UAttributes.id}")
| String
a| Unique identifier for the message

* MUST be set to the https://www.rfc-editor.org/rfc/rfc4122.html#section-3[hyphenated string representation] of the UUID.

| UserProperty
| ("2", "{UAttributes.type}")
| String
a| The type of message

> * MUST be the enum value for UAttributes.type link:../up-core-api/uprotocol/uattributes.proto[UMessageType enum].

Reviewer 3: ```suggestion
* MUST be the enum value for UAttributes.type link:../up-core-api/uprotocol/v1/uattributes.proto[UMessageType enum].
```

Author: updated

| UserProperty
| ("3","{UAttributes.source}")
| String
a| The origin (address) of the message

* MUST be the link:../basics/uri.adoc[string serialization of the UUri]

| UserProperty
| ("4","{UAttributes.sink}")
| String
a| The destination (address) of the message

* MUST be the link:../basics/uri.adoc[string serialization of the UUri]

| UserProperty
| ("5","{UAttributes.priority}")
| String
a| The message's priority as defined in link:../basics/qos.adoc[QoS doc]

> * MUST be set to the enum value for UAttributes.priority link:../up-core-api/uprotocol/uattributes.proto[UPriority enum].

Reviewer 3: ```suggestion
* MUST be set to the enum value for UAttributes.priority link:../up-core-api/uprotocol/v1/uattributes.proto[UPriority enum].
```

Author: updated

| UserProperty
| ("6", "{UAttributes.ttl}")
| String
a| The amount of time (in milliseconds) after which this message MUST NOT be delivered/processed anymore

| UserProperty
| ("7", "{UAttributes.permissionLevel}")
| String
a| The service consumer's permission level as defined in link:../up-l2/permissions.adoc#_code_based_access_permissions_caps[Code-Based uE Access Permissions (CAPs)]

| UserProperty
| ("8", "{UAttributes.commStatus}")
| String
a| A UCode indicating an error that has occurred during the delivery of either an RPC Request or Response message

> * MUST be set to the link:../up-core-api/uprotocol/ustatus.proto[UCode]'s integer value as a string.

Reviewer 3: ```suggestion
* MUST be set to the link:../up-core-api/uprotocol/v1/ustatus.proto[UCode]'s integer value as a string.
```

Author: updated

| UserProperty
| ("9", "{UAttributes.reqId}")
| String
a| The identifier that a service consumer can use to correlate an RPC Repsonse message with its RPC Request

* MUST be set to the https://www.rfc-editor.org/rfc/rfc4122.html#section-3[hyphenated string representation] of the UUID.

| UserProperty
| ("10", "{UAttributes.token}")
| String
a| The service consumer's access token

| UserProperty
| ("11", "{UAttributes.traceParent}")
| String
a| A tracing identifier to use for correlating messages across the system

| UserProperty
| ("12", "{UAttributes.payload_format}")
| String
a| The format for the data stored in the UMessage

> * MUST be set to the enum value for UAttributes.payload_format link:../up-core-api/uprotocol/uattributes.proto[UPayloadFormat enum].

Reviewer 3: ```suggestion
* MUST be set to the enum value for UAttributes.payload_format link:../up-core-api/uprotocol/v1/uattributes.proto[UPayloadFormat enum].
```

Author: updated
